activesupport/lib/active_support/encryptor/jwe.rb

```diff
@@ -0,0 +1,157 @@
+require 'base64'
+require 'openssl'
+require 'active_support/core_ext/hash/keys'
+
+module ActiveSupport
+  class MessageEncryptor
+    class JWE # :nodoc:
+      attr_reader :cipher
+
+      def initialize(**options)
+        @options = options
+        @secret = @options[:secret]
+        @alg = key_encryption_algorithm
+        @cipher = encryption_algorithm
+        @enc = CIPHER_ENC_PAIRS[@cipher]
+        @serializer = @options[:serializer] || Marshal
+      end
+
+      def encrypt(value)
+        protected_header = @serializer.dump header
+
+        cipher = new_cipher
+        cipher.encrypt
+
+        mac_key, cipher.key = derive_mac_and_enc_key(content_encryption_key)
+        cipher.iv = iv = cipher.random_iv
+        auth_data = encode(protected_header)
+        if gcm?
+          cipher.auth_data = auth_data
+          ciphertext = cipher.update(@serializer.dump value) + cipher.final
+          auth_tag = cipher.auth_tag
+        else
+          ciphertext = cipher.update(@serializer.dump value) + cipher.final
+          auth_data_length = [auth_data.length * 8].pack('Q>')
+          mac_input = [auth_data, iv, ciphertext, auth_data_length].join
+          mac = generate_digest(mac_input, mac_key)
+          auth_tag = mac[0...(mac.length / 2)]
+        end
+
+        auth_data << '.' << [jwe_encrypted_key, iv, ciphertext, auth_tag].map { |a| encode a }.join('.')
+      end
+
+      def decrypt(encrypted_message)
+        parts = encrypted_message.split('.')
+        auth_data = parts.shift
+        encrypted_key, iv, ciphertext, auth_tag = parts.map { |a| decode a }
```

Comment:
Hi there! Looks like we've been doing similar work (#25874) 😄 you'll want to check the auth_tag length here and add a test for it.